keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,494 @@
+
+The end result of those gradual changes in terms of the format and content of logs and the actual API functionalities could be very similar to the ones which will be produced by the proposed logr-based API. The main difference between the two approaches would lay in the path of the migration.
+
+The main advantage of using klog compatible API for the whole migration for structured logs  would be the possibility of introducing changes to the API and log format in a gradual and less disturbing way and still maintaining the possibility of producing backward compatible logs (why this is important is discussed in the next section of this document).
+
+However at the same time there are some issues which are harder to address using this approach:
+* In order to have fully queryable stable logs a few changes which might require some manual work or verification from developers needs to be introduced. Those include introduction of constant log message which could be used to identify log entries, extraction of all relevant log entry key-values/fields, passing context between functions or adding Kubernetes objects to this context. Although those changes can be introduced gradually in a one-by-one fashion the fact that each of those will require involvement of potentially large number of developers means that those changes will most likely be much harder to perform this way than they would be using the all-in-one approach.
+* Current klog API is quite messy in some places. This includes things like having multiple logging levels (info/warning/error/fatal) which quite often cannot be clearly separated (info vs. warning, warning vs. error) and the large number of variants of the same method (Info(), Infof(), Infoln(), InfoDepth()). Those things make using logging API harder and it would be good to address those. Switching to a new API might make this migration easier.   
+
+#### Avoid double writing of logs using klog and logr APIs
+
+Instead of leaving the existing references to klog (or actually a klog wrapper) for an extended period of time until the legacy output format is completely removed from the Kubernetes and having to do a double writing to both klog and logr APIs for all existing logs we could consider two alternative approaches:
+* Use logr-based API for all existing and new logs.
+* Introduce intermediate API which would be a superset of the logr API and could produce exactly the same logs as the existing klog API.
+
+The main problem with the first approach is the fact that it makes very hard to produce log lines which would be exactly the same as the existing klog-based logs. This could be done for instance by adding special fields in logr-based logs called `legacy-message`, `legacy-level` and `legacy-source` which could be lazy evaluated and which could be used only by the logr legacy format.
```

Comment:
Is that really that big problem? We will remove them at some point anyway. So we are just moving this problem by couple quarters...

Reply:
Removing all usages of one API (klog) is probably a bit easier to do then removing some specific arguments to the given API (logr) but the real problem with this approach is the maintenance of the compatibility when it comes to the content of log entries in the legacy format. Without an API guarding from accidental changes it will be easy to break this compatibility.